src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -2700,6 +2655,146 @@ DlLockDataType DoorLockServer::credentialTypeToLockDataType(DlCredentialType cre
+        // grant access if it is not provided.
+        bool requirePin = false;
+        auto err        = Attributes::RequirePINforRemoteOperation::Get(endpoint, &requirePin);
+
+        VerifyOrExit(
+            EMBER_ZCL_STATUS_SUCCESS == err,
+            emberAfDoorLockClusterPrintln("Failed to read 'RequirePINforRemoteOperations' attribute: status=0x%" PRIx8, err));
+        credentialsOk = !requirePin;
+    }
+
+    // TODO: increase WrongCodeEntryLimit if credentialsOk == false.
+    // TODO: If limit is exceeded, lock remote operations for UserCodeTemporaryDisableTime.
+    VerifyOrExit(credentialsOk, {
+        reason = DlOperationError::kInvalidCredential;
+        emberAfDoorLockClusterPrintln("Checking credentials failed: either PIN is invalid or not provided");
+    });
+
+    // credentials check succeeded, try to lock/unlock door
+    operationOk = opHandler(endpoint, pinCode, reason);
+    VerifyOrExit(operationOk, /* reason is set by the above call */);
+
+    // door locked, set cluster attribute
+    VerifyOrDie(SetLockState(endpoint, newLockState, DlOperationSource::kRemote));
+
+exit:
+    bool success = credentialsOk && operationOk;
+
+    // Send command response
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Send LockOperation/LockOperationError event
+    LockOpCredentials foundCred[] = { { DlCredentialType::kPin, pinCredIdx } };
+    LockOpCredentials * credList  = nullptr;
+    size_t credListSize           = 0;
+
+    // appclusters.pdf 5.3.5.3, 5.3.5.4:
+    // The list of credentials used in performing the lock operation. This SHALL be null if no credentials were involved.
+    if (pinCode.HasValue())
+    {
+        credList     = foundCred;
+        credListSize = 1;
+    }
+
+    SendLockOperationEvent(endpoint, opType, DlOperationSource::kRemote, reason, Nullable<uint16_t>(pinUserIdx),
+                           Nullable<chip::FabricIndex>(getFabricIndex(commandObj)), Nullable<chip::NodeId>(getNodeId(commandObj)),
+                           credList, credListSize, success);
+    return success;
+}
+
+void DoorLockServer::SendLockOperationEvent(chip::EndpointId endpointId, DlLockOperationType opType, DlOperationSource opSource,
+                                            DlOperationError opErr, Nullable<uint16_t> userId,
+                                            Nullable<chip::FabricIndex> fabricIdx, Nullable<chip::NodeId> nodeId,
+                                            LockOpCredentials * credList, size_t credListSize, bool opSuccess)
+{
+    Nullable<List<const Structs::DlCredential::Type>> credentials{};
+
+    // appclusters.pdf 5.3.5.3, 5.3.5.4:
+    // The list of credentials used in performing the lock operation. This SHALL be null if no credentials were involved.
+    if (nullptr == credList || 0 == credListSize)
+    {
+        credentials.SetNull();
```

Comment:
I don't think that this is required, `Nullables` default constructor makes them `null` anyway.